sled-agent/src/instance.rs

```diff
@@ -854,10 +854,81 @@ impl InstanceRunner {
         running_zone: &RunningZone,
         migrate: Option<InstanceMigrationTargetParams>,
     ) -> Result<(), Error> {
-        let nics = running_zone
+        use propolis_client::{
+            types::{
+                BlobStorageBackend, Board, BootOrderEntry, BootSettings,
+                Chipset, ComponentV0, CrucibleStorageBackend,
+                InstanceInitializationMethod, NvmeDisk, QemuPvpanic,
+                ReplacementComponent, SerialPort, SerialPortNumber, VirtioDisk,
+                VirtioNetworkBackend, VirtioNic,
+            },
+            PciPath, SpecKey,
+        };
+
+        struct DiskComponents {
+            id: Uuid,
+            device: NvmeDisk,
+            backend: CrucibleStorageBackend,
+        }
+
+        struct NicComponents {
+            id: Uuid,
+            device: VirtioNic,
+            backend: VirtioNetworkBackend,
+        }
+
+        // Construct the lists of disks and NICs that will need to be sent to
+        // Propolis.
+        //
+        // N.B. When initializing a VM via live migration, the caller may
+        // replace the descriptions of some kinds of backend components. This is
+        // used here to pass new Crucible VCRs and OPTE port names to the new
+        // Propolis. This replacement requires that the source and target agree
+        // on the IDs of the backend components being replaced. To provide this
+        // stability, identify disk and NIC backends using the corresponding
+        // disk and interface UUIDs.
+        let disks: Vec<DiskComponents> = self
+            .requested_disks
+            .iter()
+            .map(|disk| -> Result<DiskComponents, std::io::Error> {
+                // NVMe disk serial numbers are supposed to be left-justified
+                // and padded with spaces (see version 1.2.1 of the spec,
+                // section 1.5). However, some versions of Propolis padded
+                // serial numbers with zeroes instead. Disk serial numbers can
```

Comment:
will we always do this? or could this be a part of some versioning that will eventually be done away with in a V2?

Reply:
Yeah, we would like to fix this eventually. I believe oxidecomputer/propolis#790 and perhaps also oxidecomputer/propolis#789 are the relevant Propolis issues; perhaps this ought to be documented someplace here?